File: backend/services/user/user_service.py
"""
User service module for handling user-related operations.
"""

import uuid
from typing import Any, Dict, Optional

from api.schemas.users import SchemaUserCreate
from api.security import get_password_hash
from db.models import User, UserPreferences
from services.common.base_service import BaseService


class UserService(BaseService):
    """Service for handling user-related operations."""

    def check_email_exists(self, email: str) -> bool:
        """
        Check if a user with the given email already exists.

        Args:
            email: The email to check

        Returns:
            bool: True if the email exists, False otherwise
        """
        return self.db.query(User).filter(User.email == email).first() is not None

    def check_handle_exists(self, handle: str) -> bool:
        """
        Check if a user with the given handle already exists.

        Args:
            handle: The handle to check

        Returns:
            bool: True if the handle exists, False otherwise
        """
        return self.db.query(User).filter(User.handle == handle).first() is not None

    def create_user(self, user_data: SchemaUserCreate) -> User:
        """
        Create a new non-guest user in the system.

        Args:
            user_data: The user data to create

        Returns:
            User: The created user

        Raises:
            ValueError: If the email or handle already exists
        """
        # Check if email already exists
        if self.check_email_exists(user_data.email):
            raise ValueError("Email already registered")

        # Check if handle already exists
        if self.check_handle_exists(user_data.handle):
            raise ValueError("Handle already taken")

        # Create hashed password
        hashed_password = get_password_hash(user_data.password)

        # Set default profile image if not provided
        profile_image = user_data.profile_image
        if not profile_image:
            profile_image = "/profile/profile.jpg"

        # Prepare user metadata
        user_metadata: Dict[str, Any] = {}

        new_user = User(
            id=uuid.uuid4(),
            email=user_data.email,
            name=user_data.name,
            handle=user_data.handle,
            password_hash=hashed_password,
            bio=user_data.bio,
            profile_image=profile_image,
            user_metadata=user_metadata,
        )

        self.db.add(new_user)
        self.db.commit()
        self.db.refresh(new_user)

        self.logger.info(f"Created new user: {new_user.name} ({new_user.email})")
        return new_user

    def get_user_by_id(self, user_id: str) -> Optional[User]:
        """
        Get a user by ID.

        Args:
            user_id: The user ID

        Returns:
            User: The user if found, None otherwise
        """
        return self.get_by_id(User, user_id)

    def get_user_by_email(self, email: str) -> Optional[User]:
        """
        Get a user by email.

        Args:
            email: The user's email

        Returns:
            User: The user if found, None otherwise
        """
        return self.db.query(User).filter(User.email == email).first()

    def get_user_by_handle(self, handle: str) -> Optional[User]:
        """
        Get a user by handle.

        Args:
            handle: The user's handle

        Returns:
            User: The user if found, None otherwise
        """
        return self.db.query(User).filter(User.handle == handle).first()

    def update_user(self, user_id: str, update_data: Dict[str, Any]) -> Optional[User]:
        """
        Update a user's information.

        Args:
            user_id: The user ID
            update_data: Data to update

        Returns:
            User: The updated user if found, None otherwise
        """
        user = self.get_user_by_id(user_id)
        if not user:
            return None

        # Update fields
        for key, value in update_data.items():
            if hasattr(user, key) and key != "id" and key != "password_hash":
                setattr(user, key, value)

        self.db.commit()
        self.db.refresh(user)
        return user

    def get_user_preferences(self, user_id: str) -> Optional[UserPreferences]:
        """
        Get user preferences.

        Args:
            user_id: The user ID

        Returns:
            UserPreferences: The user preferences if found, None otherwise
        """
        return self.db.query(UserPreferences).filter(UserPreferences.user_id == user_id).first()

    def update_user_preferences(self, user_id: str, preferences_data: Dict[str, Any]) -> Optional[UserPreferences]:
        """
        Update user preferences.

        Args:
            user_id: The user ID
            preferences_data: Preferences data to update

        Returns:
            UserPreferences: The updated preferences if found, None otherwise
        """
        preferences = self.get_user_preferences(user_id)

        if not preferences:
            # Create new preferences if not found
            preferences = UserPreferences(id=uuid.uuid4(), user_id=user_id, **preferences_data)
            self.db.add(preferences)
        else:
            # Update existing preferences
            for key, value in preferences_data.items():
                if hasattr(preferences, key) and key != "id" and key != "user_id":
                    setattr(preferences, key, value)

        self.db.commit()
        self.db.refresh(preferences)
        return preferences

File: frontend/app/services/userService.ts
/**
 * User service for handling user-related operations
 * Updated to use Axios with Zustand auth store
 */

import { ApiUserPreferencesSchema, ApiUserSchema } from '@/app/types/schema';
import logger from '../lib/logger';
import { useAuthStore } from '../store/authStore';
import { apiClient } from './apiMiddleware';
import { buildParams } from './utils';

/**
 * Interface for API results
 */
export interface ApiResult<T> {
  data?: T;
  error?: string;
  status: number;
  originalError?: any;
}

/**
 * Interface for user update data
 * @see components.schemas.UserUpdate from api-types.ts
 */
export interface UserUpdateData {
  name?: string;
  email?: string;
  handle?: string;
  bio?: string;
  profileImage?: string;
  password?: string;
}

/**
 * Interface for user preferences update data
 * @see components.schemas.UserPreferencesUpdate from api-types.ts
 */
export interface UserPreferencesUpdateData {
  theme?: 'light' | 'dark' | 'system';
  emailNotifications?: boolean;
  pushNotifications?: boolean;
  timeZone?: string;
  language?: string;
  additionalSettings?: Record;
}

/**
 * Helper to process API responses consistently
 */
function processApiResponse<T>(response: any): ApiResult {
  return {
    data: response.data,
    status: response.status,
  };
}

/**
 * Helper to handle API errors consistently
 */
function handleApiError<T>(error: any): ApiResult {
  if (error.response) {
    // Server responded with an error status
    return {
      error: error.response.data?.detail || 'Server error',
      status: error.response.status,
      originalError: error,
    };
  } else if (error.request) {
    // Request was made but no response received
    return {
      error: 'No response from server',
      status: 0,
      originalError: error,
    };
  } else {
    // Something else went wrong
    return {
      error: error.message || 'Unknown error',
      status: 0,
      originalError: error,
    };
  }
}

/**
 * Service for user-related operations using Axios with Zustand
 */
export class UserService {
  /**
   * Get current user profile
   * First tries to get from Zustand store for efficiency
   * @returns Promise with API result containing user data
   */
  static async getCurrentUser(): Promise {
    try {
      // Check if we already have user data in the store
      const currentUser = useAuthStore.getState().currentUser;

      // If we have a current user in the store and no force refresh,
      // return it without making an API call to avoid loops
      if (currentUser) {
        return {
          data: currentUser as ApiUserSchema,
          status: 200,
        };
      }

      // IMPORTANT: If no token exists, don't make the API call
      if (typeof window !== 'undefined') {
        const token = localStorage.getItem('auth_token');
        if (!token) {
          return {
            error: 'No authentication token available',
            status: 401,
          };
        }
      }

      // If no data in store but we have a token, make API call
      // Use a direct axios call to avoid triggering interceptors that might cause loops

      // Using the API types directly
      const response = await apiClient.get<ApiUserSchema>('/users/me');

      // Log the user profile data for debugging
      logger.debug(`[Auth] /users/me API response: ${JSON.stringify(response.data, null, 2)}`);

      // Update store with user data
      if (response.data) {
        // Use setState to avoid triggering getters/methods
        useAuthStore.setState({ currentUser: response.data });
        logger.debug(`[Auth] Updated currentUser in store`);
      }

      return processApiResponse<ApiUserSchema>(response);
    } catch (error) {
      return handleApiError<ApiUserSchema>(error);
    }
  }

  /**
   * Get a user by ID
   * @param userId - User ID
   * @returns Promise with API result containing user data
   */
  static async getUserById(userId: string): Promise {
    try {
      const response = await apiClient.get<ApiUserSchema>(`/users/${userId}`);
      return processApiResponse<ApiUserSchema>(response);
    } catch (error) {
      return handleApiError<ApiUserSchema>(error);
    }
  }

  /**
   * Get a user by handle
   * @param handle - User handle
   * @returns Promise with API result containing user data
   */
  static async getUserByHandle(handle: string): Promise {
    try {
      const response = await apiClient.get<ApiUserSchema>(`/users/handle/${handle}`);
      return processApiResponse<ApiUserSchema>(response);
    } catch (error) {
      return handleApiError<ApiUserSchema>(error);
    }
  }

  /**
   * Search for users
   * @param query - Search query (optional)
   * @param limit - Maximum number of results (default: 20)
   * @param offset - Pagination offset (default: 0)
   * @returns Promise with API result containing user data array
   */
  static async searchUsers(query?: string, limit: number = 20, offset: number = 0): Promise {
    try {
      // Build query params
      const params = buildParams({ query, limit, offset });

      const response = await apiClient.get<ApiUserSchema[]>('/users', { params });
      return processApiResponse<ApiUserSchema[]>(response);
    } catch (error) {
      return handleApiError<ApiUserSchema[]>(error);
    }
  }

  /**
   * Update current user profile
   * @param userData - User data to update
   * @returns Promise with API result containing updated user data
   */
  static async updateCurrentUser(userData: UserUpdateData): Promise {
    try {
      const response = await apiClient.put<ApiUserSchema>('/users/me', userData);

      // Update user in store if update was successful
      if (response.data) {
        useAuthStore.getState().setCurrentUser(response.data);
      }

      return processApiResponse<ApiUserSchema>(response);
    } catch (error) {
      return handleApiError<ApiUserSchema>(error);
    }
  }

  /**
   * Get current user preferences
   * @returns Promise with API result containing user preferences
   */
  static async getUserPreferences(): Promise {
    try {
      const response = await apiClient.get<ApiUserPreferencesSchema>('/users/me/preferences');
      return processApiResponse<ApiUserPreferencesSchema>(response);
    } catch (error) {
      return handleApiError<ApiUserPreferencesSchema>(error);
    }
  }

  /**
   * Update user preferences
   * @param preferencesData - Preferences data to update
   * @returns Promise with API result containing updated preferences
   */
  static async updateUserPreferences(preferencesData: UserPreferencesUpdateData): Promise {
    try {
      const response = await apiClient.put<ApiUserPreferencesSchema>('/users/me/preferences', preferencesData);
      return processApiResponse<ApiUserPreferencesSchema>(response);
    } catch (error) {
      return handleApiError<ApiUserPreferencesSchema>(error);
    }
  }

  /**
   * Request password reset
   * @param email - User email
   * @returns Promise with API result
   */
  static async requestPasswordReset(email: string): Promise {
    try {
      const response = await apiClient.post('/users/reset-password/request', { email });
      return processApiResponse<{ message: string }>(response);
    } catch (error) {
      return handleApiError<{ message: string }>(error);
    }
  }

  /**
   * Confirm password reset
   * @param token - Reset token
   * @param newPassword - New password
   * @returns Promise with API result
   */
  static async confirmPasswordReset(token: string, newPassword: string): Promise {
    try {
      const response = await apiClient.post('/users/reset-password/confirm', {
        token,
        newPassword, // Using camelCase
      });
      return processApiResponse<{ message: string }>(response);
    } catch (error) {
      return handleApiError<{ message: string }>(error);
    }
  }

  /**
   * Upload profile image
   * @param file - Image file to upload
   * @returns Promise with API result containing URL of uploaded image
   */
  static async uploadProfileImage(file: File): Promise {
    try {
      const formData = new FormData();
      formData.append('file', file);

      // For demo purposes, our backend upload URL might be at media/upload/profile
      const response = await apiClient.post('/media/upload/profile', formData, {
        headers: {
          'Content-Type': 'multipart/form-data',
        },
      });

      // Log the response for debugging
      console.log('Profile image upload response:', response.data);

      return processApiResponse<{ url: string }>(response);
    } catch (error) {
      console.error('Profile image upload error:', error);
      return handleApiError<{ url: string }>(error);
    }
  }
}
